signature/src/balrog.rs

```diff
@@ -0,0 +1,363 @@
+    let pem_chain = parse_pem_chain(x5u)?;
+    let balrog = Balrog::new(&pem_chain)?;
+
+    /* Verify things. */
+    let _ = balrog.verify_chain(current_time, root_hash)?;
+    let _ = balrog.verify_leaf_hostname(leaf_subject)?;
+    let _ = balrog.verify_content_signature(input, signature)?;
+
+    /* Success */
+    Ok(())
+}
+
+impl<'a> Balrog<'_> {
+    pub fn new(list: &'a Vec<Pem>) -> Result<Balrog<'a>, BalrogError> {
+        if list.is_empty() {
+            return Err(BalrogError::CertificateNotFound);
+        }
+
+        let mut parsed: Vec<X509Certificate<'a>> = Vec::new();
+        let mut hash: Vec<u8> = Vec::new();
+        for (i, pem) in list.iter().enumerate() {
+            if pem.label != "CERTIFICATE" {
+                return Err(BalrogError::from(X509Error::InvalidCertificate));
+            }
+
+            let x509 = pem.parse_x509()?;
+            parsed.push(x509);
+
+            /* For the final certificate, calculate its SHA256 hash. */
+            /* TODO: Technically, we can get more bytes in the PEM beyond the end
+             * of the ASN.1 certificate sequence, and this function will include
+             * them in the hash. This feels like an unclear edge case though.
+             */
+            if i + 1 == list.len() {
+                let digest = digest::digest(&digest::SHA256, pem.contents.as_slice());
+                hash.extend_from_slice(digest.as_ref());
+            }
+        }
+
+        Ok(Balrog {
+            chain: parsed,
+            root_hash: hash,
+        })
+    }
+
+    /* Ensure that the certificate chain is valid. */
+    pub fn verify_chain(&self, current_time: i64, root_hash: &str) -> Result<(), BalrogError> {
+        if self.chain.is_empty() {
+            return Err(BalrogError::CertificateNotFound);
+        }
+
+        /* For each certificate N in the chain, check that:
+         *  1. The issuer name in cert N matches the subject name of cert N+1.
+         *  2. The validity time of cert N contains the current time.
+         *  3. The cert N+1 contains a key usage extension permitting CA.
+         *  4. Extract the public key from cert N+1
+         *  5. Call X509Certificate::verify_signature with the key from step 4.
+         *
+         * For the last certificate in the chain (the root), repeat the same
+         * steps using the last certificate as both cert N and cert N+1, and
+         * check that the SHA256 of the root certificate matches the root_hash
+         * parameter to this function.
+         */
+        let mut index = 0;
+        while index + 1 < self.chain.len() {
+            let subject = &self.chain[index];
+            let issuer = &self.chain[index + 1];
+            let _ = Self::verify_cert_chain_pair(subject, issuer, current_time)?;
+            index = index + 1;
+        }
+
+        /* Verify the root certificate. */
+        let root_cert = match self.chain.last() {
+            None => return Err(BalrogError::CertificateNotFound),
```

Comment:
The `None` case here occurs when `self.chain` is empty, and this replaces the assert that used to exist for this condition. This should only be able to happen if the caller instantiated the `Balrog` struct without invoking `Balrog::new()` or somehow manages to skip parsing the certificate chain.

Reply:
Ah, ok. Is https://github.com/mozilla-mobile/mozilla-vpn-client/pull/8182/files/21ef66cf3f40d7256612f7328dc9fe96104e2a4b#diff-4e66d35def1809107b8f16c1c551fb923cc0466677ca97466c994cdac7ae9872R139-R141 not enough though? If it is enough we can just `unwrap` or `expect` here, since it's almost impossible that it will be `None`.

Reply:
Oh right, we are effectively checking for an empty chain twice.